File: app/src/main/java/me/bkkn/ui/list/NoteAdapter.java
package me.bkkn.ui.list;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import me.bkkn.domain.entity.Note;

public class NoteAdapter extends RecyclerView.Adapter<NoteViewHolder> {
    private List<Note> data = new ArrayList<>();
    private List<Note> dataCopy = new ArrayList<>();
    private NoteViewHolder.OnNoteListener onNoteListener;

    public void setOnDeleteClickListener(NoteViewHolder.OnNoteListener onNoteListener) {
        this.onNoteListener = onNoteListener;
    }

    public void setData(List<Note> notes) {
        data = notes;
        data.addAll(data);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public NoteViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final LayoutInflater inflater = (LayoutInflater) parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        return new NoteViewHolder(inflater, parent, onNoteListener);
    }

    @Override
    public void onBindViewHolder(@NonNull NoteViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    private Note getItem(int position) {
        return data.get(position);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void filter(String query) {
        dataCopy.addAll(data);
        data.clear();
        if(query.isEmpty()){
            data.addAll(dataCopy);
        } else{
            query = query.toLowerCase();
            for(Note item: dataCopy){
                if(item.getTitle().toLowerCase().contains(query) || item.getText().toLowerCase().contains(query)){
                    data.add(item);
                }
            }
        }
        notifyDataSetChanged();
    }
}

File: app/src/main/java/me/bkkn/ui/list/NoteAdapter.kt
package me.bkkn.ui.list

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import me.bkkn.domain.entity.Note
import me.bkkn.ui.list.NoteViewHolder.OnNoteListener

class NoteAdapter : RecyclerView.Adapter<NoteViewHolder>() {
    private var data: MutableList<Note> = mutableListOf()
    private lateinit var onNoteListener: OnNoteListener
    fun setOnDeleteClickListener(onNoteListener: OnNoteListener) {
        this.onNoteListener = onNoteListener
    }

    fun setData(notes: MutableList<Note>) {
        data = notes
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NoteViewHolder {
        val inflater =
            parent.context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        return NoteViewHolder(inflater, parent, onNoteListener)
    }

    override fun onBindViewHolder(holder: NoteViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    private fun getItem(position: Int): Note {
        return data[position]
    }

    override fun getItemCount(): Int {
        return data.size
    }
}